File: android/src/main/java/com/rebeloid/unity_ads/UnityAdsListener.java
package com.rebeloid.unity_ads;

import com.unity3d.ads.IUnityAdsListener;
import com.unity3d.ads.UnityAds;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodChannel;

public class UnityAdsListener implements IUnityAdsListener {
    private final Map<String, MethodChannel> placementChannels = new HashMap<>();
    private final BinaryMessenger messenger;
    private final MethodChannel defaultChannel;

    public UnityAdsListener(MethodChannel defaultChannel, BinaryMessenger messenger) {
        this.defaultChannel = defaultChannel;
        this.messenger = messenger;
    }

    @Override
    public void onUnityAdsReady(String placementId) {
        onReady(placementId);
    }

    @Override
    public void onUnityAdsStart(String placementId) {
        onStarted(placementId);
    }

    @Override
    public void onUnityAdsFinish(String placementId, UnityAds.FinishState result) {
        switch (result) {
            case SKIPPED:
                onSkipped(placementId);
                break;
            case COMPLETED:
                onCompleted(placementId);
                break;
            case ERROR:
                onError(placementId, null, null);
                break;
        }
    }

    @Override
    public void onUnityAdsError(UnityAds.UnityAdsError error, String message) {
        onError(null, error, message);
    }

    private MethodChannel findChannel(String placementId) {
        if (placementChannels.containsKey(placementId)) {
            return placementChannels.get(placementId);
        }
        MethodChannel methodChannel = new MethodChannel(messenger,
                UnityAdsConstants.VIDEO_AD_CHANNEL + "_" + placementId);
        placementChannels.put(placementId, methodChannel);
        return methodChannel;
    }

    private void onReady(String placementId) {
        HashMap<String, Object> arguments = new HashMap<>();
        arguments.put(UnityAdsConstants.PLACEMENT_ID_PARAMETER, placementId);

        invokeMethod(UnityAdsConstants.READY_METHOD, arguments, placementId);
    }

    private void onStarted(String placementId) {
        HashMap<String, Object> arguments = new HashMap<>();
        arguments.put(UnityAdsConstants.PLACEMENT_ID_PARAMETER, placementId);

        invokeMethod(UnityAdsConstants.START_METHOD, arguments, placementId);
    }

    private void onCompleted(String placementId) {
        HashMap<String, Object> arguments = new HashMap<>();
        arguments.put(UnityAdsConstants.PLACEMENT_ID_PARAMETER, placementId);

        invokeMethod(UnityAdsConstants.COMPLETE_METHOD, arguments, placementId);
    }

    private void onSkipped(String placementId) {
        HashMap<String, Object> arguments = new HashMap<>();
        arguments.put(UnityAdsConstants.PLACEMENT_ID_PARAMETER, placementId);

        invokeMethod(UnityAdsConstants.SKIPPED_METHOD, arguments, placementId);
    }

    private void onError(String placementId, UnityAds.UnityAdsError error, String errorMessage) {
        Map<String, Object> arguments = new HashMap<>();

        if (placementId != null) {
            arguments.put(UnityAdsConstants.PLACEMENT_ID_PARAMETER, placementId);
        }

        if (error != null) {
            arguments.put("errorCode", error.toString());
        }

        if (errorMessage != null) {
            arguments.put("errorMessage", errorMessage);
        }

        invokeMethod(UnityAdsConstants.ERROR_METHOD, arguments, placementId);
    }

    private void invokeMethod(String methodName, Map<String, Object> arguments, String placementId) {
        defaultChannel.invokeMethod(methodName, arguments);
        if (placementId != null) {
            findChannel(placementId).invokeMethod(methodName, arguments);
        }
    }
}

File: ios/Classes/UnityAdsListener.swift
import Flutter
import UnityAds

public class UnityAdsListener : NSObject, UnityAdsDelegate {
    
    var placementChannels: [String: FlutterMethodChannel] = [:]
    
    let messenger: FlutterBinaryMessenger
    let defaultChannel: FlutterMethodChannel
    
    init(messenger: FlutterBinaryMessenger, defaultChannel: FlutterMethodChannel) {
        self.messenger = messenger
        self.defaultChannel = defaultChannel
        super.init()
    }
    
    public func unityAdsReady(_ placementId: String) {
        invokeMethod(UnityAdsConstants.READY_METHOD,
                     args: [UnityAdsConstants.PLACEMENT_ID_PARAMETER: placementId],
                     placementId: placementId)
    }
    
    public func unityAdsDidError(_ error: UnityAdsError, withMessage message: String) {
        onError(error: error, errorMessage: message)
    }
    
    public func unityAdsDidStart(_ placementId: String) {
        invokeMethod(UnityAdsConstants.START_METHOD,
                     args: [UnityAdsConstants.PLACEMENT_ID_PARAMETER: placementId],
                     placementId: placementId)
    }
    
    public func unityAdsDidFinish(_ placementId: String, with state: UnityAdsFinishState) {
        switch (state) {
            case UnityAdsFinishState.skipped:
                onSkipped(placementId);
                break;
            case UnityAdsFinishState.completed:
                onCompleted(placementId);
                break;
            case UnityAdsFinishState.error:
                onError(placementId: placementId);
                break;
            default:
                return
        }
    }
    
    func onCompleted(_ placementId:String) {
        invokeMethod(UnityAdsConstants.COMPLETE_METHOD,
                     args: [UnityAdsConstants.PLACEMENT_ID_PARAMETER: placementId],
                     placementId: placementId)
    }
    
    func onSkipped(_ placementId:String) {
        invokeMethod(UnityAdsConstants.SKIPPED_METHOD,
                     args: [UnityAdsConstants.PLACEMENT_ID_PARAMETER: placementId],
                     placementId: placementId)
    }
    
    func onError(placementId:String! = nil, error:UnityAdsError! = nil,  errorMessage:String! = nil) {
        var arguments: [String:String] = [:]
        
        if (placementId != nil) {
            arguments[UnityAdsConstants.PLACEMENT_ID_PARAMETER]=placementId
        }
        
        if (error != nil) {
            arguments["errorCode"]=String(describing:  error)
        }
        
        if (errorMessage != nil) {
            arguments["errorMessage"]=errorMessage
        }
        
        invokeMethod(UnityAdsConstants.ERROR_METHOD, args: arguments, placementId: placementId)
    }
    
    func invokeMethod(_ methodName: String, args: [String:String],placementId: String?){
        defaultChannel.invokeMethod(methodName, arguments: args)
        if (placementId != nil) {
            var channel = placementChannels[placementId!]
            if (channel == nil) {
                channel = FlutterMethodChannel(name: UnityAdsConstants.VIDEO_AD_CHANNEL + "_" + placementId!, binaryMessenger: messenger)
                placementChannels[placementId!] = channel
            }
            channel?.invokeMethod(methodName, arguments: args)
        }
    }
}
